services/src/main/java/org/keycloak/services/HttpResponseImpl.java

```diff
@@ -44,6 +56,31 @@ public void setHeader(String name, String value) {
         delegate.getOutputHeaders().putSingle(name, value);
     }
 
+    @Override
+    public void addCookieIfAbsent(HttpCookie cookie) {
+        if (cookie == null) {
+            throw new IllegalArgumentException("Cookie is null");
+        }
+
+        if (cookies == null) {
+            cookies = new HashSet<>();
+        }
+
+        if (cookies.add(cookie)) {
+            if (writeCookiesOnTransactionComplete) {
```

Comment:
There's a potential problem here if `writeCookiesOnTransactionComplete` is enabled after a cookie is set in the first place, which would result in the cookie being added one more time. Maybe just theoretical though.

Reply:
I mean, this call should always happen at the very beginning of the request. That is why encapsulating this is important to make sure the situation you mentioned does not happen.

It does not make sense to enable this flag after cookies are already set.

Reply:
It's probably fine as it is now, and we can discuss any follow-ups around it in combination with discussing the retry stuff in general.